### ISSUE TITLE
[camera] Add support for web

### ISSUE THREAD

**Issue by hterkelsen** [flutter/flutter#45297]
We should add web support to the `camera` plugin

**Comment by incube8r** [comment]
How about? https://medium.com/@mk.pyts/how-to-access-webcam-video-stream-in-flutter-for-web-1bdc74f2e9c7

**Comment by ditman** [comment]
This rewrite: https://github.com/flutter/flutter/issues/31225 might affect the decisions we need to take to implement the plugin. 

Please, follow up in that ticket before starting work on camera for web.

**Comment by LautaroLobo12** [comment]
Is there any release date for this? At least, this year?

**Comment by ditman** [comment]
@LautaroLobo12 yes, at least taking photos could come this year. Recording videos might be more challenging.

**Comment by rahulkushwaha482** [comment]
@pedromassangocode  please add this features in web . It's urgent.

**Comment by Hexer10** [comment]
I've made some progress here implementing the camera APIs on web: https://github.com/Hexer10/plugins/tree/camera-web/packages/camera/camera_web .

Most of the APIs are not yet implemented, tough it is already possible to preview the camera and take the pictures.

**Comment by ditman** [comment]
@Hexer10 and @felangel should talk in a few weeks :)

**Comment by 7flash** [comment]
> I've made some progress here implementing the camera APIs on web: https://github.com/Hexer10/plugins/tree/camera-web/packages/camera/camera_web .
> 
> Most of the APIs are not yet implemented, tough it is already possible to preview the camera and take the pictures.
> 
> EDIT:
> Implemented zoom, video recoding (tough the downloaded file does not show the video length and it is only available in `video/webm` format.)

Thank you, that sounds great. Can you please tell how we can use your version of the plugin?

**Comment by c-antin** [comment]
What about https://www.github.com/flutter/photobooth/tree/main/packages%2Fcamera%2Fcamera_web by Very Good Ventures: https://medium.com/flutter/how-its-made-i-o-photo-booth-3b8355d35883

**Comment by nolandg** [comment]
Just switching from React to Flutter and trying to get camera working in web platform. The Photobooth project @c-antin metioned seems great https://github.com/flutter/flutter/issues/45297#issuecomment-850805518. 

What is the Flutter/Dart way of using the Photobooth `camera` package in my project? Is there a better way than just pasting that folder into my project? With JS npm I could specificy the git addresses and commit hash in my `package.json` to import a module and keep it up-to-date with an open source project. Does Flutter/Dart have the same concept? Can I use the Photobooth `camera` package directly in my project? It seems they have not published it to pub.dev

Thanks!

**Comment by ditman** [comment]
@nolandg the VGV team is working on creating the web version of the `camera` plugin, so at the end of the day you can specify a "package: camera" dependency on your Flutter app, and it supports mobile + web more or less automatically. Take a look at the [url_launcher](https://pub.dev/packages/url_launcher) for a plugin that supports mmmmmany platforms!

**Comment by nolandg** [comment]
Thanks @ditman, that's great to hear. I was just itching to use it now don't want to be another one of those "is it done yet?" comments :-)

**Comment by ditman** [comment]
@nolandg no problem, the camera web plugin is the latest hotness right now. There's some users banding to put something together here, I tried to offer them a path forward (reusing as much code from the photobooth as possible):

https://github.com/flutter/plugins/pull/3868#issuecomment-875834800

**Comment by daraul** [comment]
Is there any (hacky, even) solution to get the camera working on web *today*?

**Comment by treeder** [comment]
It's certainly possible without using the `camera` interface, we did it here for QR code scanning: https://github.com/treeder/jsqr_flutter/ Can dig through the scanner.dart code for how.

**Comment by ditman** [comment]
@daraul, absolutely, look at https://github.com/flutter/photobooth (the code in that package is being used as the basis for the camera plugin of the web)

**Comment by Hexer10** [comment]
> Is there any (hacky, even) solution to get the camera working on web _today_?

@daraul If you want to try out my implementation from this pr ( https://github.com/flutter/plugins/pull/3868 ) you can check a demo here https://hexer10.github.io/camera_test_web/  ( https://github.com/Hexer10/camera_test_web ). It works implementing the already in use `camera` interface.

**Comment by mrtdeh** [comment]
Will this feature added until next month?

**Comment by ditman** [comment]
@mrtdeh it's being actively worked on, I can't give you specific dates because I don't know them.

**Comment by Snowy-27** [comment]
i finally succed 
https://github.com/Snowy-27/cameraflutterweb

**Comment by ditman** [comment]
@Snowy-27 why not use the `camera_web` package that is being worked on from a Git dart dependency? I think the plugin is ready for that use case.

**Comment by ditman** [comment]
Hello all!

We've just published one of the latest PRs for the first version of the camera_web package (we still have a PR for video recording pending, but it'll get merged soon), so we've decided to gather some feedback from the community!

Here's the example app, compiled and deployed for the web (woohoo!):

https://dit-vgv-camera-test.web.app

And this is how you can modify your pubspec.yaml to get the package in your apps (at least until it's published in pub.dev):

```
  camera_web:
    git:
      url: https://github.com/flutter/plugins.git
      ref: master
      path: packages/camera/camera_web
```

For those who're interested in the camera plugin for the web, it'd be great if you could give this "preview" version a shot, and report back with your impressions/bug reports/PRs/words of encouragement/words of disgust/anything. (PS: if you create a new issue, remember to mention me there so I'm aware of it!)

In about a week, @bselwe and I will look at all the feedback and then decide if we publish this as a 0.8.0 version (almost there) or a 0.3.0 version (needs some more cooking!) 😉

Interesting links:

* Docs in the [README of the camera_web plugin](https://github.com/flutter/plugins/blob/master/packages/camera/camera_web/README.md).
* Look at the [changes to the camera example app](https://github.com/flutter/plugins/pull/4240) to see what you need to do to make your app web-friendly.

Thanks!

**Comment by scognito** [comment]
> Hello all!
> 
> We've just published one of the latest PRs for the first version of the camera_web package (we still have a PR for video recording pending, but it'll get merged soon), so we've decided to gather some feedback from the community!
> 
> Here's the example app, compiled and deployed for the web (woohoo!):
> 
> https://dit-vgv-camera-test.web.app
> 
> This is what you need to add to your pubspec.yaml to get the package in your apps (at least until it's published in pub.dev):
> 
> ```
>   camera: ^...
>   camera_web:
>     git:
>       url: https://github.com/flutter/plugins.git
>       ref: master
>       path: packages/camera/camera_web
> ```
> 
> For those who're interested in the camera plugin for the web, it'd be great if you could give this "preview" version a shot, and report back with your impressions/bug reports/PRs/words of encouragement/words of disgust/anything. (PS: if you create a new issue, remember to mention me there so I'm aware of it!)
> 
> In about a week, @bselwe and I will look at all the feedback and then decide if we publish this as a 0.8.0 version (almost there) or a 0.3.0 version (needs some more cooking!) 😉
> 
> Interesting links:
> 
> * Docs in the [README of the camera_web plugin](https://github.com/flutter/plugins/blob/master/packages/camera/camera_web/README.md).
> * Look at the [changes to the camera example app](https://github.com/flutter/plugins/pull/4240) to see what you need to do to make your app web-friendly.
> 
> Thanks!

Awesome job, we as community need a camera plugin for web!
I tried the demo on my Xiaomi 10 lite 5g, here the result:
* App starts correctly (nice start 🙂)
* I had to mute using the speaker button because of classic echo you listen when you have 2 devices with mic open on both
* Preview widget is 16:9 landscape but the picture is taken in portrait (even with unlock rotation option).

Hope it helps the development, keep up the good work!

**Comment by daraul** [comment]
Does this use the [stream API](https://caniuse.com/stream)? I find myself working with some old devices lately; they don't support it.

**Comment by ditman** [comment]
> I tried the demo on my Xiaomi 10 lite 5g, here the result:
> 
> * App starts correctly (nice start 🙂)
> * I had to mute using the speaker button because of classic echo you listen when you have 2 devices with mic open on both
> * Preview widget is 16:9 landscape but the picture is taken in portrait (even with unlock rotation option).

Thanks for the testing! I can see the problems you're mentioning in my own testing too, so it's not device/browser specific! We'll keep an eye on those!

**Comment by ditman** [comment]
> Does this use the [stream API](https://caniuse.com/stream)? I find myself working with some old devices lately; they don't support it.

@daraul yes, It does, it is documented in the README of the web plugin, [here](https://github.com/flutter/plugins/blob/master/packages/camera/camera_web/README.md).

What older devices are you needing to support? Have you found any fallbacks for the Stream API?

**Comment by daraul** [comment]
> What older devices are you needing to support?

A couple of professional terminals running rather old versions of android ([1] [2] [3]).

> Have you found any fallbacks for the Stream API?

Not yet. I am considering either installing opera mini on all devices, or doing some heavy googling. I am fairly certain this kind of thing was possible on older versions of android, but I could be wrong. 

[1]: http://www.td-tech.com/en/product/index.aspx?nodeid=115&page=ContentPage&contentid=714
[2]: http://www.td-tech.com/en/product/index.aspx?nodeid=115&page=ContentPage&contentid=698
[3]: http://www.td-tech.com/en/product/index.aspx?nodeid=115&page=ContentPage&contentid=314

**Comment by bselwe** [comment]
Hey @Xyzcancer, which version of Chrome are you using on your device? Is that possible that another Chrome tab is using your camera? If so, please try closing it.

**Comment by Xyzcancer** [comment]
@bselwe thank you for quick reply! I've requested info and will provide it to you. But I'm sure there is no other tab in Chrome with camera. Also I want to be sure that I'm using it right, I'm a bit confused.
I have two dependencies:
  camera: ^0.9.0
  camera_web:
    git:
      url: https://github.com/flutter/plugins.git
      ref: master
      path: packages/camera/camera_web
      
 So, there is no import of camera web in example. Is camera plugin is using it internally or I just misunderstood how to integrate it?

**Comment by bselwe** [comment]
Yeah, that is correct usage. The plugin is not published to pub.dev yet.

**Comment by Xyzcancer** [comment]
@bselwe okay, I see. Version of chrome mobile is 92.0.4515.159

**Comment by Xyzcancer** [comment]
@bselwe how's your investigation going? By the way, I've tried to get photo directly by ImageCapture like this:
var stream = _webcamVideoElement.srcObject as MediaStream;
var blob = await ImageCapture(stream.getVideoTracks().last).takePhoto();

And method takePhoto() returns Blob which I can't use as file. Could you please give me a hint here?

**Comment by bselwe** [comment]
@Xyzcancer Your browser should support loading the camera, so it might be a device error. Are you able to run this example app on your device correctly: https://webrtc.github.io/samples/src/content/getusermedia/gum/?

As for the `takePhoto` you might need to display the image with `Image.network` rather than `Image.file` ([see here why](https://github.com/flutter/plugins/blob/master/packages/camera/camera_web/README.md#taking-a-picture)).

Did you manage to get access to camera directly without using the plugin? If so, could you share a sample?

**Comment by Xyzcancer** [comment]
Yes, I did.
_webcamVideoElement = VideoElement();
    // Register an webcam
    ui.platformViewRegistry.registerViewFactory(
        'webcamVideoElement', (int viewId) => _webcamVideoElement);
    // Create video widget
    _webcamWidget =
        HtmlElementView(key: UniqueKey(), viewType: 'webcamVideoElement');
    // Access the webcam stream
    window.navigator.getUserMedia(video: true).then((MediaStream stream) {
      print('getUserMediaCalled');
      _webcamVideoElement.srcObject = stream;
      _webcamVideoElement.play();
    });

This code give me an access even on those devices that the plugin not able to get

**Comment by Xyzcancer** [comment]
Also I found how to convert Blob into xFile in your plugin.

var stream = _webcamVideoElement.srcObject as MediaStream;
var blob = await ImageCapture(stream.getVideoTracks().last).takePhoto();
var xFile = XFile(html.Url.createObjectUrlFromBlob(blob));

**Comment by Xyzcancer** [comment]
> @Xyzcancer Your browser should support loading the camera, so it might be a device error. Are you able to run this example app on your device correctly: https://webrtc.github.io/samples/src/content/getusermedia/gum/?
> 
> As for the `takePhoto` you might need to display the image with `Image.network` rather than `Image.file` ([see here why](https://github.com/flutter/plugins/blob/master/packages/camera/camera_web/README.md#taking-a-picture)).
> 
> Did you manage to get access to camera directly without using the plugin? If so, could you share a code sample?

Yeah, webrtc link is working on those devices

**Comment by ditman** [comment]
> [This code](https://github.com/flutter/flutter/issues/45297#issuecomment-909113035) give me an access even on those devices that the plugin not able to get

@bselwe maybe it's an issue with camera constraints being too strict for that device? In the example only `video: true` seems to be in effect.

**Comment by Tidus123746** [comment]
Hi @bselwe 

I did some testing with your app,  it works great on my computer in the browser,  but when I try to run it on my phone, it is unable to find my camera.  This is using both the Safari and chrome browser on Android using a galaxy S20+.

Is this plugin expected to work on phone browsers aswell, or is this just meant to work in a desktop environment?

Thanks

**Comment by bselwe** [comment]
Thanks @Tidus123746, @Xyzcancer for your feedback. 

@Tidus123746, are you receiving a `cameraNotReadable` error when using the [example app](https://dit-vgv-camera-test.web.app/)?

I have tested this on Galaxy S21 and had the same issue as well. Fortunately, I found a way to fix this, so it's going to be addressed soon!

**Comment by Tidus123746** [comment]
Hi @bselwe 

I'm not sure exactly the error, I don't know where I would see that,  but I just get the "no camera found" text at the bottom where you would select it.  But if you had a similar issue on the S21, I'm sure its probably the same underlying error.  That's great you found a way to resolve it,  and awesome work on this branch aswell, it is very exciting to be able to use the camera on web. 

Thanks

**Comment by ditman** [comment]
We've published the first iteration of the `camera_web` package:

https://pub.dev/packages/camera_web

There's some pending PRs (like recording video, which is currently under some thorough testing), and maybe outstanding bugs.

I'm going to close this old issue as fixed. If you come across an issue with the package, have suggestions, etc... please [**do create a new issue**](https://github.com/flutter/flutter/issues/new/choose), as you'd do with any other plugin!

Thanks!
